### PR TITLE
Add Username, Password, Database and Unix Socket support.

### PR DESCRIPTION
**Pull request by Vedant-Gandhi**
## Description
The aim of this PR is to add support for including the Username, Password, Database and Unix Socket Path for Redis.

## Type
- New Feature

## Reference Issue
- https://github.com/rtCamp/nginx-helper/issues/350
- https://github.com/rtCamp/nginx-helper/issues/343

## Screenshots
Screenshot 2024-09-06 at 12 20 46

## Testing
The code has been tested locally and is working as expected.

### CODE REVIEW

File: admin/class-phpredis-purger.php [hunk 1]
```diff
@@ -39,11 +39,32 @@ public function __construct() {
...
+			$password = $nginx_helper_admin->options['redis_password'];
+			
+			if( $username && $password ) {
+				$redis_acl['auth'] = array( $username, $password );
+			}
+			
+			$hostname = empty( $nginx_helper_admin->options['redis_unix_socket'] ) ? $nginx_helper_admin->options['redis_hostname'] : $nginx_helper_admin->options['redis_unix_socket'];
+			$port     = empty( $nginx_helper_admin->options['redis_unix_socket'] ) ? $nginx_helper_admin->options['redis_port'] : 0;
+			
 			$this->redis_object->connect(
```
Comment (SH4LIN) [inline admin/class-phpredis-purger.php:55]: On my local setup, there’s no Nginx server or Redis installed. When I switch branches from master to your branch, the code throws a fatal error. The plugin is trying to connect, but the connection is being refused. If the plugin is attempting to make a connection, there should also be a fatal error in the master branch.

Please take a look into this

File: admin/partials/nginx-helper-general-options.php [hunk 2]
```diff
@@ -22,6 +22,12 @@
 	'redis_hostname',
 	'redis_port',
 	'redis_prefix',
+    'redis_database',
```
Comment (SH4LIN) [inline admin/partials/nginx-helper-general-options.php:25]: Fix the indentation

File: admin/partials/nginx-helper-general-options.php [hunk 3]
```diff
@@ -271,8 +284,32 @@
...
+									
+									echo '<p class="description">';
+									esc_html_e( 'Overridden by unix socket path.', 'nginx-helper' );
+									echo '</p>';
+									
+								}
+								?>
 							</td>
 						</tr>
+                        <tr>
```
Comment (SH4LIN) [inline admin/partials/nginx-helper-general-options.php:298]: Fix the indentation and use tabs instead of the space

**Review by SH4LIN — CHANGES_REQUESTED** [review]

File: admin/class-phpredis-purger.php (hunk 1 above)
Comment (Vedant-Gandhi) [reply admin/class-phpredis-purger.php:55]: @SH4LIN I am unable to replicate the issue on my end. Can you please share the error here if possible?

Comment (SH4LIN) [reply admin/class-phpredis-purger.php:55]: Sure.

Comment (SH4LIN) [reply admin/class-phpredis-purger.php:55]: We are throwing a fatal error when the Redis connection is refused, which is understandable. However, this makes the entire admin side inaccessible. As a result, there is no way to update the connection configuration except by manually updating the value in the database or using WP-CLI.

One solution is to wrap the `$this->redis_object->connect` inside try catch and on the catch display WP notice

image

Comment (Vedant-Gandhi) [reply admin/class-phpredis-purger.php:55]: @SH4LIN I found where the issue was and have fixed it. In predis it does not have checks for connection and instead immeditely executes the statement without checking if a connection is already established or not.

So in class-predis.php where we have addded selecting the databse feature, it was selecting the database without checking for connection establishment causing it to throw error.

**Comment by SH4LIN** [comment]
@Vedant-Gandhi Is this ready for review?

**Comment by Vedant-Gandhi** [comment]
@SH4LIN Yes

**Review by SH4LIN — APPROVED** [review]

**Pull request merged**